File: Dynamic Array and Ring buffer/lib/ring_buffer.rb
require_relative "static_array"

class RingBuffer
  attr_reader :length

  def initialize
    @store = StaticArray.new(0)
    @length = 0
    @capacity = 8
    @start_idx = @capacity - 1
  end

  # O(1)
  def [](index)
    check_index(index)
    @store[(index + @start_idx + 1) % @capacity]
  end

  # O(1)
  def []=(index, value)
    check_index(index)
    @store[(index + @start_idx + 1) % @capacity] = value
  end

  # O(1)
  def pop
    check_index
    val = self[@length - 1]
    self[@length - 1] = nil
    @length -= 1
    val
  end

  # O(1) ammortized
  def push(val)
    resize! if @length == @capacity
    @length += 1
    self[@length - 1] = val
  end

  # O(1)
  def shift
    check_index
    val = self[0]
    self[0] = nil
    @length -= 1
    @start_idx = (@start_idx + 1) % @capacity
    val
  end

  # O(1) ammortized
  def unshift(val)
    resize! if @length == @capacity
    @length += 1
    @store[@start_idx] = val
    @start_idx = (@start_idx - 1) % @capacity
  end

  protected
  attr_accessor :capacity, :start_idx, :store
  attr_writer :length

  def check_index(index = 0)
    raise 'index out of bounds' if @length.zero? || !index.between?(0, @length - 1)
  end

  def resize!
    new_store = StaticArray.new(@capacity)
    (0...@length).each do |i|
      new_store[i] = self[i]
    end

    @capacity *= 2
    @start_idx = @capacity - 1
    @store = new_store
  end
end

File: Algorithms class-Coursera/Data Structure/RingBuffer.js
class RingBuffer {
  constructor() {
    this.capacity = 8;
    this.store = new Array(this.capacity);
    this.length = 0;
    this.startIdx = 0;
  }

  get(i) {
    // O(1)
    this.inBoundValidation(i);
    i = (this.startIdx + i) % this.capacity;
    return this.store[i];
  }

  set(i, val) {
    this.inBoundValidation(i);
    i = (this.startIdx + i) % this.capacity;
    if (this.store[i] === undefined) {
      // set new element => addition
      this.ensureLengthAndResize();
      this.length++;
    }
    if (val === undefined) {
      // set element to undefined => deletion
      this.length--;
    }
    this.store[i] = val;
  }

  push(val) {
    // O(1) ammortized
    this.ensureLengthAndResize();
    this.set(this.length, val);
  }

  unshift(val) {
    // O(1)
    this.ensureLengthAndResize();
    this.startIdx = (this.startIdx - 1) % this.capacity;
    if (this.startIdx < 0) { this.startIdx = this.capacity + this.startIdx; }
    this.set(0, val);
  }

  shift() {
    // O(1)
    this.checkEmpty();
    let returnEl = this.get(0);
    this.set(0, undefined);
    this.startIdx = (this.startIdx + 1) % this.capacity;
    if (this.startIdx < 0) { this.startIdx = this.capacity + this.startIdx; }
    return returnEl;
  }

  pop() {
    // O(1)
    this.checkEmpty();
    let returnEl = this.get(this.length - 1);
    this.set(this.length - 1, undefined);
    return returnEl;
  }

  checkEmpty() {
    if (this.length === 0) {
      throw 'Store is empty.';
    }
  }

  ensureLengthAndResize() {
    // O(n) to resize the array
    if (this.length === this.capacity) {
      let newStore = new Array(this.capacity * 2);
      for (let i=0; i<this.capacity; i++) {
        newStore[i] = this.get(i);
      }
      this.capacity *= 2;
      this.store = newStore;
      this.startIdx = 0;
    }
  }

  inBoundValidation(i) {
    if (i < 0 || i >= this.capacity) {
      throw 'Index is not in Bound';
    }
  }
}

module.exports = RingBuffer;
